[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/billydavila/CSC448/blob/main/Judging%20Flowers/Judging_flowers.ipynb)


# Judging Flowers (Iris Dataset)

![irisPic](https://raw.githubusercontent.com/billydavila/CSC448/main/Judging%20Flowers/irisPic.png)

# Imports and pip installations (if needed)

In [1]:
# pip installations
!pip install tabulate


# libraries
from sklearn.datasets import load_iris

from sklearn import linear_model, model_selection, preprocessing, pipeline, metrics, svm, neural_network, neighbors

import pandas as pd
import numpy as np

from tabulate import tabulate
import matplotlib.pyplot as plt
import warnings

warnings.filterwarnings("ignore")

***
# Part 1: Load the dataset

In [2]:
# Load the dataset (load remotely, not locally)

# load the data
iris = load_iris() 

# Display to console
print( "Data: ") 
print( iris.data[:10])
print( "\nFeature Names:", end = "  " )
print( iris.feature_names )
print( "\nTarget: ", end = " ")
print(iris.target)
print("\nTarget Names: ", end = " ")
print(iris.target_names)

Data: 
[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]]

Feature Names:  ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']

Target:  [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]

Target Names:  ['setosa' 'versicolor' 'virginica']


In [3]:
# Put everything nicely in a single DataFrame
featureDataFrame = pd.DataFrame( data = iris.data, columns = iris.feature_names, )
listTarget_names = [f'{iris.target_names[0]}']*50 + [f'{iris.target_names[1]}']*50 + [f'{iris.target_names[2]}']*50
speciesDataFrame = pd.DataFrame( data = listTarget_names, columns = ['species'])
targetDataFrame = pd.DataFrame( data = iris.target, columns = ['target'] )
irisDataFrame = pd.concat( [featureDataFrame, speciesDataFrame,targetDataFrame ], axis = 1)

# Output the first 15 rows of the data
irisDataFrame.head(15)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4                1.6               0.2   
12                4.8               3.0                1.4               0.1   
13                4.3               3.0                1.1               0.1   
14                5.8               4.0                1.2               0.2   

   species  target  
0   setosa       0  
1   setosa       0  
2   setosa       0  
3   setosa       0  
4   setosa       0  
5   setosa       0  
6   setosa       0  
7   setosa       0  
8   setosa       0  
9   setosa       0  
10  setosa       0  
11  setosa       0  
12  setosa       0  
13  setosa       0  
14  setosa       0

In [4]:
# Display a summary of the table information (number of datapoints, etc.)
irisDataFrame.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   species            150 non-null    object 
 5   target             150 non-null    int32  
dtypes: float64(4), int32(1), object(1)
memory usage: 6.6+ KB


In [5]:
irisDataFrame.describe() # Display the Descriptive statistics 

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)      target  
count        150.000000  150.000000  
mean           1.199333    1.000000  
std            0.762238    0.819232  
min            0.100000    0.000000  
25%            0.300000    0.000000  
50%            1.300000    1.000000  
75%            1.800000    2.000000  
max            2.500000    2.000000

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?
![irisDataPictures](https://raw.githubusercontent.com/billydavila/CSC448/main/Judging%20Flowers/iris-datasetPic.png)


* The iris data set is widely used as a beginner's dataset for machine learning purposes. The dataset is included in R base and Python in the machine learning package Scikit-learn, so that users can access it without having to find a source for it. Several versions of the dataset have been published (But we will be using the one provided by `sklearn.datasets`).
* The dataset consists of 50 samples from each of the three species of Iris (Iris setosa, Iris virginica and Iris versicolor). Four features were measured from each sample: the length and the width of the sepal and petals, in centimeters.

![irisExplainedPic](https://raw.githubusercontent.com/billydavila/CSC448/main/Judging%20Flowers/IrisDataSet.png)


* The dataset contains a set of 150 records under five attributes - sepal length, sepal width, petal length, petal width and species.
* The features are the measurements: sepal length, sepal width, petal length, petal width. 
* The labels will be the species which are mapped with numbers as: 
          0 := 'setosa' 
          1 := 'versicolor'
          2 := 'virginica'
          
* Framed as a supervised learning problem: Predict the species (0: setosa, 1: versicolor, 2: virginica) of an iris using the measurements (sepal length, sepal width, petal legth, petal width)

***
# Part 2: Split the dataset into train and test

In [6]:
# Take the dataset and split it into our features (X) and label (y)
X = irisDataFrame[iris.feature_names] # Features
Y = irisDataFrame['target']

In [7]:
X.head(3)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2

In [8]:
Y.head(3)

0    0
1    0
2    0
Name: target, dtype: int32

In [9]:
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, Y_train, Y_test = model_selection.train_test_split( X, Y, train_size = 0.9, test_size = 0.1,
                                                                     shuffle = True, random_state = 6174)

In [10]:
# Check the shapes
print ( X_train.shape, X_test.shape )
print( Y_train.shape, Y_test.shape )

(135, 4) (15, 4)
(135,) (15,)


***
# Part 3: Logistic Regression
![LogisticPic](https://raw.githubusercontent.com/billydavila/CSC448/main/Judging%20Flowers/linear_vs_logistic_regression.jpg)

In [11]:
# i. Use sklearn to train a LogisticRegression model on the training set
logistic_regression = linear_model.LogisticRegression().fit(X_train, Y_train) 
logistic_regression

LogisticRegression()

In [12]:
# ii. For a sample datapoint, predict the probabilities for each possible class

sampleDataPoint = X[:1] # Pick a sample datapoint
myPrediction = logistic_regression.predict_proba(sampleDataPoint)[0]  
# Display to console
print("Sample Data point:")
print( sampleDataPoint.to_string(index = False) )

probData = [ ['0', 'setosa', "{:0.5f}".format(myPrediction[0]) ], 
             ['1', 'versicolor', "{:0.5f}".format(myPrediction[1]) ], 
             ['2', 'virginica', "{:0.5f}".format(myPrediction[2]) ],
             ['Sum', '', myPrediction.sum()]
           ]

print( "\nProbabilities for each possible class:")
print(tabulate(probData, headers=["Class","Class name (iris type)", "Probability"], tablefmt='fancy_grid'))    

Sample Data point:
 sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
               5.1               3.5                1.4               0.2

Probabilities for each possible class:
╒═════════╤══════════════════════════╤═══════════════╕
│ Class   │ Class name (iris type)   │   Probability │
╞═════════╪══════════════════════════╪═══════════════╡
│ 0       │ setosa                   │       0.98048 │
├─────────┼──────────────────────────┼───────────────┤
│ 1       │ versicolor               │       0.01952 │
├─────────┼──────────────────────────┼───────────────┤
│ 2       │ virginica                │       0       │
├─────────┼──────────────────────────┼───────────────┤
│ Sum     │                          │       1       │
╘═════════╧══════════════════════════╧═══════════════╛


**Now let's see how well the model (Logistic Regression) predicts on the test data**

In [13]:
# Predicting the Test set results
Y_pred_log_regression = logistic_regression.predict(X_test)

# Predict probabilities
probs_y_log_regression = logistic_regression.predict_proba(X_test)

probs_y_log_regression = np.round(probs_y_log_regression, 4)

res = "{:<10} | {:<10} | {:<10} | {:<13} | {:<5}".format("y_test", "y_pred", "Setosa(%)", "versicolor(%)", "virginica(%)\n")
res += "-"*65+"\n"
res += "\n".join("{:<10} | {:<10} | {:<10} | {:<13} | {:<10}".format(x, y, a, b, c) for x, y, a, b, c in zip(Y_test, Y_pred_log_regression, 
                                                                                                             probs_y_log_regression[:,0],
                                                                                                             probs_y_log_regression[:,1], 
                                                                                                             probs_y_log_regression[:,2]))
res += "\n"+"-"*65+"\n"
print(res)

y_test     | y_pred     | Setosa(%)  | versicolor(%) | virginica(%)
-----------------------------------------------------------------
2          | 2          | 0.0001     | 0.1259        | 0.874     
1          | 1          | 0.0026     | 0.8621        | 0.1352    
2          | 2          | 0.0        | 0.0054        | 0.9946    
1          | 1          | 0.0055     | 0.8173        | 0.1772    
2          | 2          | 0.0001     | 0.0988        | 0.9012    
1          | 1          | 0.0345     | 0.9494        | 0.0161    
0          | 0          | 0.9701     | 0.0299        | 0.0       
2          | 2          | 0.0        | 0.0068        | 0.9932    
1          | 1          | 0.0655     | 0.9267        | 0.0078    
0          | 0          | 0.9501     | 0.0499        | 0.0       
0          | 0          | 0.975      | 0.025         | 0.0       
1          | 1          | 0.0086     | 0.9349        | 0.0565    
1          | 1          | 0.0198     | 0.9587        | 0.0215    
1       

**From the above table, we can see that we are correct in predicting the class all the time. Hence, the score on the test data will be 1. However, let's calculate it.** 

In [14]:
# iii. Report on the score for Logistic regression model, what does the score measure?

train_score = logistic_regression.score( X_train, Y_train) # Calculate the score from the training data set
test_score  = logistic_regression.score(X_test, Y_test)     # Calculate the score from the test data set

# Display to console
print("The training score of model is: ", train_score)
print( "The test score of model is: ", test_score )

The training score of model is:  0.9703703703703703
The test score of model is:  1.0


####  Report on the score for Logistic regression model, what does the score measure? 
* From the documentation, we know that the [score method](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html) return the mean accuracy on the given test data and labels. Hence, the score measures the accuracy of the model on the given test data and labels.
* The model got a perfect score of 1.0 on the test data. Meaning that the model accurately predicted the correct answer on the test data (as we saw previously in the above table).
* The model got a score of 0.9703 on the training data. Meaning that the model accurately predicted the correct answer most of the time on the training data.
* From the score, we can conclude that the model overall performed well on classifying iris class to which class belongs to based on the measurements (sepal's width, sepal's length, petal's width, petal's length). 

In [15]:
# iv. Extract the coefficents and intercepts for the boundary line(s)

coefficients = logistic_regression.coef_      # extract coefficents 
intercepts = logistic_regression.intercept_   # extract intercepts

# Display to console
print("Intercepts:  ", end = '')
for i in range( len(intercepts)):
    if i != len(intercepts) -1:
        print( "{:0.5f}".format(intercepts[i]), end = ", " )
    else:
        print( "{:0.5f}".format(intercepts[i]), end = "\n" )
        
print( "\nCoefficients:  ", end = '' )
for i in range( len(coefficients)):
    for j in  range( len(coefficients[i])):
        if i == len(coefficients)-1 and j == len(coefficients[i])-1:
            print( "{:0.5f}".format(coefficients[i][j]), end = "\n")
        else:
            print( "{:0.5f}".format(coefficients[i][j]), end = ", " )

Intercepts:  9.61346, 2.10967, -11.72313

Coefficients:  -0.39669, 0.90496, -2.43823, -1.08051, 0.49734, -0.28243, -0.19224, -0.88393, -0.10065, -0.62253, 2.63047, 1.96444


***
# Part 4: Support Vector Machine

<figure>
<img src="https://raw.githubusercontent.com/billydavila/CSC448/main/Judging%20Flowers/SVC.jpg" alt="Trulli" style="width:40%">

</figure>

In [16]:
# i. Use sklearn to train a Support Vector Classifier on the training set
svmClassifier = svm.SVC( kernel = 'linear', probability = True)
svmClassifier.fit( X_train, Y_train)

SVC(kernel='linear', probability=True)

In [17]:
# ii. For a sample datapoint, predict the probabilities for each possible class

sampleDataPoint = X[100:101] # Pick a sample datapoint
myPrediction = svmClassifier.predict_proba(sampleDataPoint)[0]  
# Display to console
print("Sample Data point:")
print( sampleDataPoint.to_string(index = False) )

probData = [ ['0', 'setosa', "{:0.5f}".format(myPrediction[0]) ], 
             ['1', 'versicolor', "{:0.5f}".format(myPrediction[1]) ], 
             ['2', 'virginica', "{:0.5f}".format(myPrediction[2]) ],
             ['Sum', '', myPrediction.sum()]
           ]

print( "\nProbabilities for each possible class:")
print(tabulate(probData, headers=["Class","Class name (iris type)", "Probability"], tablefmt='fancy_grid'))    

Sample Data point:
 sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
               6.3               3.3                6.0               2.5

Probabilities for each possible class:
╒═════════╤══════════════════════════╤═══════════════╕
│ Class   │ Class name (iris type)   │   Probability │
╞═════════╪══════════════════════════╪═══════════════╡
│ 0       │ setosa                   │        0.0025 │
├─────────┼──────────────────────────┼───────────────┤
│ 1       │ versicolor               │        0.0005 │
├─────────┼──────────────────────────┼───────────────┤
│ 2       │ virginica                │        0.997  │
├─────────┼──────────────────────────┼───────────────┤
│ Sum     │                          │        1      │
╘═════════╧══════════════════════════╧═══════════════╛


**Now let's see how well the model (Support Vector Machine classification) predicts on the test data**

In [18]:
# Predicting the Test set results
Y_pred_SVC = svmClassifier.predict(X_test)

# Predict probabilities
probs_y_SVC = svmClassifier.predict_proba(X_test)

probs_y_SVC = np.round(probs_y_SVC, 4)

res = "{:<10} | {:<10} | {:<10} | {:<13} | {:<5}".format("y_test", "y_pred", "Setosa(%)", "versicolor(%)", "virginica(%)\n")
res += "-"*65+"\n"
res += "\n".join("{:<10} | {:<10} | {:<10} | {:<13} | {:<10}".format(x, y, a, b, c) for x, y, a, b, c in zip(Y_test, Y_pred_SVC, 
                                                                                                             probs_y_SVC[:,0],
                                                                                                             probs_y_SVC[:,1], 
                                                                                                             probs_y_SVC[:,2]))
res += "\n"+"-"*65+"\n"
print(res)

y_test     | y_pred     | Setosa(%)  | versicolor(%) | virginica(%)
-----------------------------------------------------------------
2          | 2          | 0.0083     | 0.0721        | 0.9196    
1          | 1          | 0.0033     | 0.9757        | 0.0209    
2          | 2          | 0.0009     | 0.0004        | 0.9986    
1          | 1          | 0.0065     | 0.9296        | 0.0639    
2          | 2          | 0.0073     | 0.0586        | 0.9342    
1          | 1          | 0.0238     | 0.9679        | 0.0083    
0          | 0          | 0.9378     | 0.0464        | 0.0158    
2          | 2          | 0.0024     | 0.002         | 0.9956    
1          | 1          | 0.0427     | 0.9452        | 0.012     
0          | 0          | 0.9187     | 0.0628        | 0.0184    
0          | 0          | 0.9632     | 0.0264        | 0.0104    
1          | 1          | 0.0074     | 0.9825        | 0.0101    
1          | 1          | 0.0174     | 0.9763        | 0.0063    
1       

In [19]:
# iii. Report on the score for the SVM, what does the score measure?

train_score = svmClassifier.score( X_train, Y_train) # Calculate the score from the training data set
test_score  = svmClassifier.score(X_test, Y_test)     # Calculate the score from the test data set

# Display to console
print("The training score of model is: ", train_score)
print( "The test score of model is: ", test_score )

The training score of model is:  0.9925925925925926
The test score of model is:  1.0


#### Report on the score for SVM, what does the score measure? 
* From the documentation, we know that the [score method](https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html) returns the mean accuracy on the given test data and labels. Hence, the score measure the accuracy of the model on the given test data and labels.
* The model got a perfect score of 1.0 on the test data. Meaning that the model accurately predicted the correct answer on the test data (as we saw previously in the above table). Comparing it with the Logistic Regression, SVM had a greater probability in predicting the correct class (it was more certain about the correct answer).
* The model got a score of 0.99259 on the training data. Meaning that the model accurately predicted the correct answer most of the time on the training data (it failed on one).
* From the score, we can conclude that the model overall performed well on classifying iris class to which class belongs to based on the measuremenrs (sepal's width, sepal's length, petal's width, petal's length).


***
# Part 5: Neural Network 


<figure>
<img src = "https://raw.githubusercontent.com/billydavila/CSC448/main/Judging%20Flowers/multilayerperceptron_network.png" alt="Trulli" style="width:40%">
</figure>




In [20]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set

neuralNetworkModel = neural_network.MLPClassifier(random_state = 1, hidden_layer_sizes = (150,), solver = 'lbfgs',
                                                  max_iter = 500, activation = 'logistic' ).fit(X_train, Y_train)

neuralNetworkModel

MLPClassifier(activation='logistic', hidden_layer_sizes=(150,), max_iter=500,
              random_state=1, solver='lbfgs')

In [21]:
# ii. For a sample datapoint, predict the probabilities for each possible class

sampleDataPoint = X[100:101] # Pick a sample datapoint

myPrediction = neuralNetworkModel.predict_proba(sampleDataPoint)[0]  

# Display to console
print("Sample Data point:")
print( sampleDataPoint.to_string(index = False) )

probData = [ ['0', 'setosa', "{:0.5f}".format(myPrediction[0]) ], 
             ['1', 'versicolor', "{:0.5f}".format(myPrediction[1]) ], 
             ['2', 'virginica', "{:0.5f}".format(myPrediction[2]) ],
             ['Sum', '', myPrediction.sum()]
           ]

print( "\nProbabilities for each possible class:")
print(tabulate(probData, headers=["Class","Class name (iris type)", "Probability"], tablefmt='fancy_grid'))  

Sample Data point:
 sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
               6.3               3.3                6.0               2.5

Probabilities for each possible class:
╒═════════╤══════════════════════════╤═══════════════╕
│ Class   │ Class name (iris type)   │   Probability │
╞═════════╪══════════════════════════╪═══════════════╡
│ 0       │ setosa                   │             0 │
├─────────┼──────────────────────────┼───────────────┤
│ 1       │ versicolor               │             0 │
├─────────┼──────────────────────────┼───────────────┤
│ 2       │ virginica                │             1 │
├─────────┼──────────────────────────┼───────────────┤
│ Sum     │                          │             1 │
╘═════════╧══════════════════════════╧═══════════════╛


**Now let's see how well the model (Multi-layer Perceptron classifier with the folowing hyperparameters: hidden_layer_size = (150,), solver = 'lbfgs', max_iter = 500, activation = 'logistic') predicts on the test data**


In [22]:
# Predicting the Test set results
Y_pred_MLP = neuralNetworkModel.predict(X_test)

# Predict probabilities
probs_y_MLP = neuralNetworkModel.predict_proba(X_test)

probs_y_MLP = np.round(probs_y_MLP, 4)

res = "{:<10} | {:<10} | {:<10} | {:<13} | {:<5}".format("y_test", "y_pred", "Setosa(%)", "versicolor(%)", "virginica(%)\n")
res += "-"*65+"\n"
res += "\n".join("{:<10} | {:<10} | {:<10} | {:<13} | {:<10}".format(x, y, a, b, c) for x, y, a, b, c in zip(Y_test, Y_pred_MLP, 
                                                                                                             probs_y_MLP[:,0],
                                                                                                             probs_y_MLP[:,1], 
                                                                                                             probs_y_MLP[:,2]))
res += "\n"+"-"*65+"\n"
print(res)

y_test     | y_pred     | Setosa(%)  | versicolor(%) | virginica(%)
-----------------------------------------------------------------
2          | 2          | 0.0        | 0.0           | 1.0       
1          | 1          | 0.0        | 1.0           | 0.0       
2          | 2          | 0.0        | 0.0           | 1.0       
1          | 1          | 0.0        | 1.0           | 0.0       
2          | 2          | 0.0        | 0.0           | 1.0       
1          | 1          | 0.0        | 1.0           | 0.0       
0          | 0          | 1.0        | 0.0           | 0.0       
2          | 2          | 0.0        | 0.0           | 1.0       
1          | 1          | 0.0        | 1.0           | 0.0       
0          | 0          | 0.9999     | 0.0001        | 0.0       
0          | 0          | 1.0        | 0.0           | 0.0       
1          | 1          | 0.0        | 1.0           | 0.0       
1          | 1          | 0.0        | 1.0           | 0.0       
1       

In [23]:
# iii. Report on the score for the Neural Network, what does the score measure?

train_score = neuralNetworkModel.score( X_train, Y_train) # Calculate the score from the training data set
test_score  = neuralNetworkModel.score(X_test, Y_test)     # Calculate the score from the test data set

# Display to console
print("The training score of model is: ", train_score)
print( "The test score of model is: ", test_score )

The training score of model is:  1.0
The test score of model is:  1.0


#### Report on the score for MLPClassifier, what does the score measure? 
* From the documentation, we know that the [score method](https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html) returns the mean accuracy on the given test data and labels. Hence, the score measure the accuracy of the model on the given test data and labels.
* The model got a perfect score of 1.0 on the test data. Meaning that the model accurately predicted the correct answer on the test data (as we saw previously in the above table). Comparing it with the Logistic Regression and SVM, the MLPClassifier had a greater probability (very close to 1.0) in predicting the correct class (the model was near 100% certain about the correct answer). 
* The model got a score of 1.0 on the training data. Meaning that the model accurately predicted the correct answer on all the training data. 
* From the score, we can conclude that the model overall performed perfectly on classifying iris class to which class belongs to based on the measuremenrs (sepal's width, sepal's length, petal's width, petal's length).

#### iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)

* The highest score gotten was 1.0 on both training and testing data. 
* The model configuration was as follow:
    * hidden_layer_size = (150,)
    * solver = 'lbfgs'
    * max_iter = 500
    * activation = 'logistic'
    

***
# Part 6: K-Nearest Neighbors

![K-NNImage](https://raw.githubusercontent.com/billydavila/CSC448/main/Judging%20Flowers/K-NN.png)

In [24]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
KNN_model = neighbors.KNeighborsClassifier( n_neighbors = 1)

# TODO: Do we 'train' the model as a whole without the training data? model.fit(X,Y)
KNN_model.fit(X_train, Y_train)

KNeighborsClassifier(n_neighbors=1)

In [25]:
# ii. For a sample datapoint, predict the probabilities for each possible class

sampleDataPoint = X[100:101] # Pick a sample datapoint

myPrediction = KNN_model.predict_proba(sampleDataPoint)[0]  

# Display to console
print("Sample Data point:")
print( sampleDataPoint.to_string(index = False) )

probData = [ ['0', 'setosa', "{:0.5f}".format(myPrediction[0]) ], 
             ['1', 'versicolor', "{:0.5f}".format(myPrediction[1]) ], 
             ['2', 'virginica', "{:0.5f}".format(myPrediction[2]) ],
             ['Sum', '', myPrediction.sum()]
           ]

print( "\nProbabilities for each possible class:")
print(tabulate(probData, headers=["Class","Class name (iris type)", "Probability"], tablefmt='fancy_grid'))  

Sample Data point:
 sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
               6.3               3.3                6.0               2.5

Probabilities for each possible class:
╒═════════╤══════════════════════════╤═══════════════╕
│ Class   │ Class name (iris type)   │   Probability │
╞═════════╪══════════════════════════╪═══════════════╡
│ 0       │ setosa                   │             0 │
├─────────┼──────────────────────────┼───────────────┤
│ 1       │ versicolor               │             0 │
├─────────┼──────────────────────────┼───────────────┤
│ 2       │ virginica                │             1 │
├─────────┼──────────────────────────┼───────────────┤
│ Sum     │                          │             1 │
╘═════════╧══════════════════════════╧═══════════════╛


**Now let's see how well the model (K-nearest neighbors classifier) predicts on the test data**

In [26]:
# Predicting the Test set results
Y_pred_KNN = KNN_model.predict(X_test)

# Predict probabilities
probs_y_KNN = KNN_model.predict_proba(X_test)

probs_y_KNN = np.round(probs_y_KNN, 4)

res = "{:<10} | {:<10} | {:<10} | {:<13} | {:<5}".format("y_test", "y_pred", "Setosa(%)", "versicolor(%)", "virginica(%)\n")
res += "-"*65+"\n"
res += "\n".join("{:<10} | {:<10} | {:<10} | {:<13} | {:<10}".format(x, y, a, b, c) for x, y, a, b, c in zip(Y_test, Y_pred_KNN, 
                                                                                                             probs_y_KNN[:,0],
                                                                                                             probs_y_KNN[:,1], 
                                                                                                             probs_y_KNN[:,2]))
res += "\n"+"-"*65+"\n"
print(res)

y_test     | y_pred     | Setosa(%)  | versicolor(%) | virginica(%)
-----------------------------------------------------------------
2          | 2          | 0.0        | 0.0           | 1.0       
1          | 1          | 0.0        | 1.0           | 0.0       
2          | 2          | 0.0        | 0.0           | 1.0       
1          | 1          | 0.0        | 1.0           | 0.0       
2          | 2          | 0.0        | 0.0           | 1.0       
1          | 1          | 0.0        | 1.0           | 0.0       
0          | 0          | 1.0        | 0.0           | 0.0       
2          | 2          | 0.0        | 0.0           | 1.0       
1          | 1          | 0.0        | 1.0           | 0.0       
0          | 0          | 1.0        | 0.0           | 0.0       
0          | 0          | 1.0        | 0.0           | 0.0       
1          | 1          | 0.0        | 1.0           | 0.0       
1          | 1          | 0.0        | 1.0           | 0.0       
1       

In [27]:
# iii. Report on the score for kNN, what does the score measure?

train_score = KNN_model.score( X_train, Y_train) # Calculate the score from the training data set
test_score  = KNN_model.score(X_test, Y_test)     # Calculate the score from the test data set

# Display to console
print("The training score of model is: ", train_score)
print( "The test score of model is: ", test_score )

The training score of model is:  1.0
The test score of model is:  1.0


#### Report on the score for KNeighborsClassifier, what does the score measure? 
* From the documentation, we know that the [score method](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html) returns the mean accuracy on the given test data and labels. Hence, the score measure the accuracy of the model on the given test data and labels.
* The model got a perfect score of 1.0 on the test data. Meaning that the model accurately predicted the correct answer on the test data (as we saw previously in the above table). Comparing it with Logistic Regression, SVM, and the MLPClassifier, the KNN-Classifier had a greater probability (1.0) in predicting the correct class (the model was 100% certain about the correct answer). 
* The model got a score of 1.0 on the training data. Meaning that the model accurately predicted the correct answer on all the training data. 
* From the score, we can conclude that the model overall performed perfectly on classifying iris class to which class belongs to based on the measuremenrs (sepal's width, sepal's length, petal's width, petal's length).

***
# Part 7: Conclusions and takeaways

**In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?**

### Summary of the models' results:

| Model                          |  Configuration                                  | Training Score | Test score |
| :-----------------------------:| :----------------------------------------------:| :-------------:| :---------:|
| Logistic Regression            |     default                                     |    0.97037     |   1.0      |
| SVM ( SVM)                     | kernel = 'linear', probability = True           |    0.9925      |   1.0      |
| Neural Network (MLPClassifier)       | hidden_layer_sizes = (150,), solver = 'lbfgs', max_iter = 500, activation = 'logistic' |     1.0        |   1.0      |  
| K-nearest Neighbors            | n_neighbors = 1                                 |     1.0        |    1.0     |



* Looking at the models' scores, all the models performed overall well on classifying iris class to their corresponding class based on the measurements (sepal's width, sepal's length, petal's width, petal's length). 

####  Which model(s) performed the best on the dataset?
* The two models that performed the best on the dataset were the **Neural Network Classifier** and **K-nearest Neighbors** with a perfect score on 1.0 (for both training and test dataset). However, in predicting the probability the **K-nearest Neighbors** model was 100% certain about the correct answer on the test data set. Another important point to mention is that unlike neural network, KNN classifier does not have an expensive training phase. 

#### why do you think that is? 
* I believe this is because KNN is an example of a nonparametric model. In nonparametric models, the number of parameters are theoritically infinite (parameters grow as data grows). 
* Neural networks are somewhat related to logistic regression. Basically, we can think of logistic regression as a one layer neural network. Hence, if we add more hidden layers, we can improve on the performance of the model (which was the case in this exercise). 
* In summary, I would first approach a classification problem with simple models first (e.g. logistic regression). However, if we are not satisfied with its performance, we can try to train a computationally more expensive neural networl, which has the advantage to learn more complex, non-linear functions.

#### Did anything surprise you about the exercise? 
* The score accuracy of all the model on the testing data. They were all perfect.
* When I was testing different configurations for KNN-model, the best one was having only one neighbors while adding more neighbors will give me worse performance. 


***
# References
1. [Scikit learn: Logistic Regression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html)
2. [Scikit learn: svm.SVC](https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html)
3. [Scikit learn: Neural Network models](https://scikit-learn.org/stable/modules/neural_networks_supervised.html)
4. [Scikit learn: MLPClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html)
5. [Scikit learn: KNeighborsClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html)